src/goto-programs/remove_java_nondet.cpp

```diff
@@ -0,0 +1,389 @@
+  const auto next_instr=std::next(instr, 1);
+  assert(next_instr->is_assign());
+  // Get the name of the LHS of the assignment
+  const auto return_identifier=
+    to_code_assign(next_instr->code).lhs().get(ID_identifier);
+
+  const auto end=std::end(instructions);
+
+  // Look for an instruction where this name is on the RHS of an assignment
+  const auto matching_assignment=std::find_if(
+      next_instr,
+      end,
+      [&return_identifier](const goto_programt::instructiont &instr)
+      {
+        return is_assignment_from(instr, return_identifier);
+      });
+
+  assert(matching_assignment!=end);
+
+  // Assume that the LHS of *this* assignment is the actual nondet variable
+  const auto &code_assign=to_code_assign(matching_assignment->code);
+  const auto nondet_var=code_assign.lhs();
+  const auto source_loc=instr->source_location;
+
+  // Erase from the nondet function call to the assignment
+  const auto after_matching_assignment=std::next(matching_assignment);
+  assert(after_matching_assignment!=end);
+  const auto after_erased=
+    instructions.erase(instr, after_matching_assignment);
+
+  // Generate nondet init code
+  code_blockt init_code;
+  gen_nondet_init(
+    nondet_var,
+    init_code,
+    symbol_table,
+    source_loc,
+    false,
+    true,
+    instr_info.is_never_null,
+    message_handler,
+    max_nondet_array_length);
+
+  // Convert this code into goto instructions
+  goto_programt new_instructions;
+  goto_convert(init_code, symbol_table, new_instructions, message_handler);
+
+  // Insert the new instructions into the instruction list
+  return insert_before(
+    instructions,
+    after_erased,
+    new_instructions.instructions);
+}
+
+/*******************************************************************\
+
+Function: remove_java_nondet
+
+  Inputs:
+    message_handler: Object which prints messages.
+    symbol_table: The list of known symbols.
+    max_nondet_array_length: The maximum length of new nondet arrays.
+    goto_program: The program to modify.
+
+ Purpose: Modify a goto_programt to replace 'nondet' library functions with
+          CBMC-native nondet expressions.
+
+\*******************************************************************/
+
+static void remove_java_nondet(
+  message_handlert &message_handler,
+  symbol_tablet &symbol_table,
+  const size_t max_nondet_array_length,
+  goto_programt &goto_program)
+{
+  // Check each instruction.
+  // `process_target` may modify the list in place, and returns the next
+  // iterator to look at.
+  for(auto instruction_iterator=std::begin(goto_program.instructions),
+        end=std::end(goto_program.instructions);
+      instruction_iterator!=end;
+      instruction_iterator=process_target(
+        message_handler,
+        symbol_table,
+        max_nondet_array_length,
+        goto_program.instructions,
+        instruction_iterator))
+  {
+    // Loop body deliberately empty.
+  }
```

Comment:
The action may delete the current instruction and insert a range of new instructions. It is not sufficient to simply iterate over the instructions (using `forall_goto_program_instructions`) because the list iterator may no longer point to a valid node in the list by the end of the loop body. Besides this, after inserting instructions, we do not need to check each of the inserted instructions. Instead, we should resume checking from after the inserted range. Therefore I believe the current method is justified.